Team a/feature branch

ChillBeanJosh opened the pull request:
Merging Team A Feature Branch Into Development

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full cannon interaction system: enter, aim (pillar oscillation), and fire with state-driven flow
  * Touch and interact variants for activating cannons
  * Runtime trajectory preview and dedicated cannon camera during interactions

* **Improvements**
  * Camera updates moved to LateUpdate for smoother responsiveness
  * Player physics/state handling refined for cannon launches, gravity application, and airborne transitions
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ChillBeanJosh committed:
CannonBall State

ChillBeanJosh committed:
camera and pointer to cannon

ChillBeanJosh committed:
Merge branch 'development' into TeamA/Feature-Branch

ChillBeanJosh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/States/Cannon/CannonBallState.cs:
```diff
@@ -0,0 +1,305 @@
+
+        private protected override void OnFixedUpdate()
+        {
+            if (_isLaunching)
+            {
+                _context.Rb.linearVelocity = _targetVelocity;
+            }
+        }
```
@coderabbitai I am preforming this check within ApplyCannonLaunch() in which it is defined to the variable float verticalVelocity. The way the logic works is the player is kinematic for the first half of the projectile arc, until they reach the peak defined by  if (verticalVelocity <= 0f) in which they will continue the projectile arc and then also be effected by rb physics. 

Would you prefer I save  float verticalVelocity outside the function, assigning its logic still in ApplyCannonLaunch(), and using it within _context.Rb.linearVelocity = verticalVelocity; ?

ChillBeanJosh replied:
@coderabbitai It should be a zero Vector since we do not want the player to be able to make any adjustments while in the middle of the launch, we do this zero Vector so that the player can still follow the calculated trajectory from ApplyCannonLaunch() and apply this zero Vector when reaching the peak or downwards (verticalVelocity <= 0f) since this is when we turn kinematic to false allowing normal rb logic to take over for the duration.

ChillBeanJosh replied:
@coderabbitai For clarification we are manually calculating the arc using the projectile motion formula to calculate the path which is defined as verticalVelocity. We are not using built in physics to preform this which is why kinematic is true for the first portion, and set to false after wards to re-enable rb collision.  With my current logic the player follows verticalVelocity completing the projectile motion arc.

ChillBeanJosh replied:
@coderabbitai I see, Option 2 seems to work the best for our case since we want to be kinematic as little as possible, We have it only be kinematic in the first half so that the player is collision free, but we can simply adjust ApplyCannonLaunch() so that after calculating verticalVelocity we can then:

- _context.Rb.isKinematic = false;
- _targetVelocity = initialVelocity;
- _targetVelocity.y = verticalVelocity;

taking it outside and before the call  if (verticalVelocity <= 0f) only leaving inside:

- if (_context.IsGrounded) _launchCompleted = true;
- return


With this we keep all logic sound purely being from the calculated projectile motion while still keeping the player's rb component active

ChillBeanJosh replied:
@coderabbitai No it is fine we can scratch the idea of being collision free during the ascent and making it  _context.Rb.isKinematic = false; at the start of the function, if need be I can adjust the behavior later if it needs to be implemented.

ChillBeanJosh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/States/Cannon/CannonBallState.cs:
```diff
@@ -0,0 +1,305 @@
+        //Used to Call State Transition:
+        private bool _launchCompleted;
+        public bool LaunchCompleted => _launchCompleted;
+
+        private Vector3 _targetVelocity;
+
+        private LineRenderer _trajectoryRenderer;
+
+
+        private protected override void OnEnter()
+        {
+            _launchTimer = 0f;
+
+            _inCannon = false;
+            _isLaunching = false;
+
+            _launchCompleted = false;
+        }
+
+        private protected override void OnExit()
+        {
+            if (_lerpRoutine != null) _context.StopCoroutine(_lerpRoutine);
+
+            _inCannon = false;
+            _isLaunching = false;
+
+            _currentCannon = null;
+
+            _context.Rb.isKinematic = false;
+        }
```
@coderabbitai This shouldnt create any issues since _isAdjustingPillar get set back to false when activating the input to launch out of the cannon. And the states will never change unless _launchCompleted = true; in which we have saftey checks within OnEnter() and will only become true when activating the input to launch out of the cannon which will result in the completion of this coroutine.

ChillBeanJosh committed:
Enable Rigidbody and set target velocity in cannon launch

Enable Rigidbody physics during cannon launch and set target velocity for projectile motion.

samsonahh commented:
@coderabbitai can you lint the .cs changes against https://github.com/StudioAspen/Aspen-2025-2026/blob/development/Conventions.md

ChillBeanJosh committed:
Naming Convention Violation Fixes

ChillBeanJosh committed:
CodeRabbit Safety Check Suggestions

ChillBeanJosh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/States/Cannon/CannonBallState.cs:
```diff
@@ -0,0 +1,319 @@
+            
+            _launchTimer += deltaTime * _currentCannon.ShotPower;
+            float time = _launchTimer;
+
+            Vector3 startPos = _currentCannon.CannonBase.position;
+
+            Vector3 forward = _currentCannon.LaunchDirection.forward;
+            Quaternion angleRot = Quaternion.AngleAxis(_currentCannon.ShotAngle, _currentCannon.CannonBase.right);
+            Vector3 launchDir = angleRot * forward;
+
+            Vector3 initialVelocity = launchDir.normalized * -_currentCannon.LaunchVelocity;
+
+            //Applying 1st Order [Velocity/Slope]:
+            Vector3 displacement = (initialVelocity * time);
+
+            //Applying 2nd Order [Acceleration/Curvature]:
+            displacement.y += (0.5f * _currentCannon.Acceleration * time * time) + _currentCannon.CurrentHeight;
+
+            //Calculate a Time Step:
+            Vector3 targetPosition = startPos + displacement;
+
+            //Move To Time Step Position:
+            _context.transform.position = targetPosition;
+
+            //Player Forward Following the Arc:
+            _context.transform.rotation = Quaternion.LookRotation(launchDir, Vector3.up);
+     
+            float verticalVelocity = initialVelocity.y + _currentCannon.Acceleration * time;
+
+            // Set the velocity for physics to continue the arc
+            _targetVelocity = initialVelocity;
+            _targetVelocity.y = verticalVelocity;
+            
+            //If We Are Past The Peak of Ark:
+            if (verticalVelocity <= 0f)
+            { 
+                //Variable to Change State:
+                if (_context.IsGrounded) _launchCompleted = true;
+                return; // Stop manually setting position
+            }
```
@coderabbitai For now we can leave the system working how I have, later on we will be revamping and improving how the base player works in which within the GameplayPlayerController we will having something similar to an overlap check function which is used for all states as a normal hit detection since currently CheckGrounded() is the only one. 

For now we dont need to make any adjustments.

LordRavenPlays committed:
Basic Pathing Started

- Need to fix intersection near the end of the level
- Needs height variation fixes
- Elements need to be added in

EliksonRod committed:
Created level scene for food fight area for level 1

samsonahh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/GameplayPlayerController.cs:
```diff
@@ -25,13 +27,21 @@ public class GameplayPlayerController : MonoBehaviour
         public SlopeSensor SlopeSensor { get; private set; }
         public bool IsGrounded { get; private set; }
 
+        public bool CannonAir { get; set; } = false;
```
I would avoid a public settable variable like this. Make a setter or a function that does more and also sets it.

samsonahh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/GameplayPlayerController.cs:
```diff
@@ -25,13 +27,21 @@ public class GameplayPlayerController : MonoBehaviour
         public StateMachine<GameplayPlayerController> StateMachine { get; private set; }
         
         [field: Header("State Machine")]
+        [field: SerializeField] public GroundSuperState GroundState { get; private set; } = new();
```
Ground and air super states already exist. Why are we remaking them?

samsonahh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/GameplayPlayerController.cs:
```diff
@@ -53,7 +63,9 @@ private void Update()
         private void FixedUpdate()
         {
-            ApplyGravity();
+            if (!CannonState.CannonBallState.IsLaunching)
```
This is anti state pattern, a better idea is to move ApplyGravity to the states that use it and exclude it from the states that don't. The context doesnt need to always ApplyGravity.

KayaK747 committed:
Level1Part1

Created new scene with player and level design partially completed

ChillBeanJosh committed:
PR Request Adjustments

- Made Mutator Function for Boolean Changes

- Adjusted ApplyGravity() to that is it not always active during base/context. Must Call _context. ApplyGravity() during state to allow.

ChillBeanJosh committed:
Merge branch 'TeamA/Feature-Branch' of https://github.com/StudioAspen/Aspen-2025-2026 into TeamA/Feature-Branch

KayaK747 committed:
scenechanges

finished last part of the spline layout for part 1 of level 1

LordRavenPlays committed:
Level1part1 (#33)

ChillBeanJosh committed:
Cannon State Revisions

- Update CannonBall.cs To Be Split Into 3 Seperate States Instead of 1 Single State.

- Implemented CheckOverlap() for hit detection other than CheckGrounded().

ChillBeanJosh committed:
Suggested Changes

- Code Rabbit Suggested Change Implementation.

ChillBeanJosh committed:
Transition Updates

ChillBeanJosh committed:
Development Ready Merge

- Split CannonBall State into 3 Clean Seperate SubStates.

ChillBeanJosh committed:
Merge branch 'development' into TeamA/Feature-Branch

ChillBeanJosh committed:
3 State CannonBall Suggestions

- Note: any suggested changes for CannonBall.cs is not needed as we are no longer using this script.

ChillBeanJosh committed:
3 State CannonBall Suggestions (#35)

ChillBeanJosh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/States/Cannon/Sub States/CannonFiredState.cs:
```diff
@@ -0,0 +1,135 @@
+            {
+                if (_context.CheckOverlap(_collisionMask, CollisionCheckRadius, out Collider hit))
+                {
+                    Debug.Log($"Cannon Launch Interrupted � Hit: {hit.name}");
```
@coderabbitai you can ignore this it already is set to the proper character.

ChillBeanJosh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/States/Cannon/CannonBallState.cs:
```diff
@@ -0,0 +1,319 @@
+        private bool _isAdjustingPillar = false;
+        public bool IsAdjustingPillar => _isAdjustingPillar;
+
+        //Used to Call State Transition:
+        private bool _launchCompleted;
+        public bool LaunchCompleted => _launchCompleted;
+
```
@coderabbitai This can be disregarded CannonBallState.cs is no longer in use.

ChillBeanJosh replied:
@coderabbitai yes I meant CannonBallState.cs

ChillBeanJosh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/States/Cannon/CannonBallState.cs:
```diff
@@ -0,0 +1,319 @@
+                _context.transform.position = Vector3.Lerp(startPos, targetPos, time);
+                _context.transform.rotation = Quaternion.Slerp(startRot, targetRot, time);
+
+                //Set The Pillar Rotation To Match That Of The Shot Angle:
+                if (cannonBall.CannonPillar != null)
+                {
+                    Vector3 forward = cannonBall.LaunchDirection.forward;
+                    Vector3 launchDir = Quaternion.AngleAxis(-cannonBall.ShotAngle, cannonBall.CannonBase.right) * forward;
+                    Quaternion targetPillarRot = Quaternion.LookRotation(launchDir, cannonBall.CannonBase.up);
+                }
```
@coderabbitai This can be disregarded CannonBallState.cs is no longer in use.

ChillBeanJosh replied:
@coderabbitai Apologies I mean "CannonBallState.cs" not "CannonBall.cs".

- "CannonBall.cs" = In Use.
- "CannonBallState.cs" = Not In Use

ChillBeanJosh committed:
Cannon Prefab Setup

Created Prefabs for:

- Cannon Object {Touch}
- Cannon Object {Interact}

ChillBeanJosh committed:
Clearing Duplicate States

ChillBeanJosh commented on charonsCorner/Assets/Scripts/Runtime/Player/Gameplay/GameplayPlayerController.cs:
```diff
@@ -53,16 +56,37 @@ private void Update()
         }
 
-        private void ApplyGravity()
+        public void ApplyGravity()
         {
             Rb.AddForce(Vector3.down * _gravityAmount, ForceMode.Acceleration);
         }
```
@coderabbitai ApplyGravity() is already public.

ChillBeanJosh replied:
@coderabbitai the anti state pattern, was in reference to while we previosly had ApplyGravity() as a private function it was applied to every state, in which in order to exclude it from other states we needed to apply state conditions inside ApplyGravity() which was the nti state pattern. Making it public and only assigning it to states that use it was the solution to that issue.,
